kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.internal.net;
+
+import java.util.Deque;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+
+import io.kroxylicious.proxy.config.VirtualCluster;
+
+public class EndpointRegistry {
+    private final Deque<NetworkBindingRequest> queue = new ConcurrentLinkedDeque<>();
+
+    private final Map<Endpoint, CompletableFuture<Integer>> endpointFuture = new ConcurrentHashMap<>();
+    private final Map<Endpoint, Map<VirtualClusterBindingKey, VirtualClusterBinding>> registry = new ConcurrentHashMap<>();
+
+    public CompletionStage<Endpoint> registerEndpoint(Endpoint key, String sniHostname, VirtualCluster cluster) {
```

Comment:
Evolving my ideas for this API:

registerVirtualCluster(virtualCluster) => CompletionStage
unregisterVirtualCluster(virtualCluster) => CompletionStage
reconcileBrokerEndpoints(Set<Integer> nodeIds) => CompletionStage

identifyVirtualCluster(String bindingAddress, int port, String sniHostname) => (VirtualCluster, nodeId)